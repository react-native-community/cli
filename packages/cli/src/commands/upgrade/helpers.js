// @flow
import https from 'https';

export const fetch = (url: string) =>
  new Promise<string>((resolve, reject) => {
    const request = https.get(url, response => {
      if (response.statusCode < 200 || response.statusCode > 299) {
        reject(
          new Error(`Failed to load page, status code: ${response.statusCode}`)
        );
      }
      const body = [];
      response.on('data', chunk => body.push(chunk));
      response.on('end', () => resolve(body.join('')));
    });
    request.on('error', err => reject(err));
  });
