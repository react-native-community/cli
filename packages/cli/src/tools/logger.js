/**
 * @flow
 */
import chalk from 'chalk';

const SEPARATOR = ', ';

const formatMessages = (messages: Array<string>) =>
  chalk.reset(messages.join(SEPARATOR));

const success = (...messages: Array<string>) => {
  console.log(`${chalk.green.bold('success')} ${formatMessages(messages)}`);
};

const info = (...messages: Array<string>) => {
  console.log(`${chalk.cyan.bold('info')} ${formatMessages(messages)}`);
};

const warn = (...messages: Array<string>) => {
  console.warn(`${chalk.yellow.bold('warn')} ${formatMessages(messages)}`);
};

const error = (...messages: Array<string>) => {
  console.error(`${chalk.red.bold('error')} ${formatMessages(messages)}`);
};

const debug = (...messages: Array<string>) => {
  console.log(`${chalk.gray.bold('debug')} ${formatMessages(messages)}`);
};

const log = (...messages: Array<string>) => {
  console.log(`${formatMessages(messages)}`);
};

export default {
  success,
  info,
  warn,
  error,
  debug,
  log,
};
