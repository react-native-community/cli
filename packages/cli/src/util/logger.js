/**
 * @flow
 */
import chalk from 'chalk';

const SEPARATOR = ', ';

const joinMessages = (messages: Array<string>) => messages.join(SEPARATOR);

const info = (...messages: Array<string>) => {
  console.log(
    `${chalk.black.bgCyan(' INFO ')} ${chalk.reset(joinMessages(messages))}`
  );
};

const warn = (...messages: Array<string>) => {
  console.warn(
    `${chalk.black.bgYellow(' WARN ')} ${chalk.yellow(joinMessages(messages))}`
  );
};

const error = (...messages: Array<string>) => {
  console.error(
    `${chalk.black.bgRed(' ERROR ')} ${chalk.red(joinMessages(messages))}`
  );
};

const debug = (...messages: Array<string>) => {
  console.log(`${chalk.black.bgWhite(' DEBUG ')} ${joinMessages(messages)}`);
};

module.exports = {
  info,
  warn,
  error,
  debug,
};
