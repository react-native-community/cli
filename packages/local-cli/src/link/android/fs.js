/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @format
 */

const fs = require('fs-extra');

exports.readFile = file => () => fs.readFileSync(file, 'utf8');

exports.writeFile = (file, content) =>
  content
    ? fs.writeFileSync(file, content, 'utf8')
    : c => fs.writeFileSync(file, c, 'utf8');
