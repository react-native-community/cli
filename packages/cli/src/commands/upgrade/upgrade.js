// @flow
import path from 'path';
import fs from 'fs';
import chalk from 'chalk';
import semver from 'semver';
import execa from 'execa';
import type {ContextT} from '../../tools/types.flow';
import logger from '../../tools/logger';
import PackageManager from '../../tools/PackageManager';
import {fetch} from './helpers';
import legacyUpgrade from './legacyUpgrade';

type FlagsT = {
  legacy: boolean,
};

const rnDiffPurgeUrl =
  'https://github.com/react-native-community/rn-diff-purge';

const getLatestRNVersion = async (): Promise<string> => {
  logger.info('No version passed. Fetching latest...');
  const {stdout} = await execa('npm', ['info', 'react-native', 'version']);
  return stdout;
};

const getRNPeerDeps = async (
  version: string,
): Promise<{[key: string]: string}> => {
  const {stdout} = await execa('npm', [
    'info',
    `react-native@${version}`,
    'peerDependencies',
    '--json',
  ]);

  return JSON.parse(stdout);
};

const getPatch = async (currentVersion, newVersion, projectDir) => {
  let patch;

  const rnDiffAppName = 'RnDiffApp';
  const {name} = require(path.join(projectDir, 'package.json'));

  logger.info(`Fetching diff between v${currentVersion} and v${newVersion}...`);

  try {
    patch = await fetch(
      `${rnDiffPurgeUrl}/compare/version/${currentVersion}...version/${newVersion}.diff`,
    );
  } catch (error) {
    logger.error(
      `Failed to fetch diff for react-native@${newVersion}. Maybe it's not released yet?`,
    );
    logger.info(
      'For available releases to diff see: https://github.com/react-native-community/rn-diff-purge#version-changes',
    );
    return null;
  }

  return patch
    .replace(new RegExp(rnDiffAppName, 'g'), name)
    .replace(new RegExp(rnDiffAppName.toLowerCase(), 'g'), name.toLowerCase());
};

const getVersionToUpgradeTo = async (argv, currentVersion, projectDir) => {
  const newVersion = argv[0]
    ? semver.valid(argv[0]) ||
      (semver.coerce(argv[0]) ? semver.coerce(argv[0]).version : null)
    : await getLatestRNVersion();

  if (!newVersion) {
    logger.error(
      `Provided version "${
        argv[0]
      }" is not allowed. Please pass a valid semver version`,
    );
    return null;
  }

  if (currentVersion > newVersion) {
    logger.error(
      `Trying to upgrade from newer version "${currentVersion}" to older "${newVersion}"`,
    );
    return null;
  }
  if (currentVersion === newVersion) {
    const {
      dependencies: {'react-native': version},
    } = require(path.join(projectDir, 'package.json'));

    if (semver.satisfies(newVersion, version)) {
      logger.warn(
        `Specified version "${newVersion}" is already installed in node_modules and it satisfies "${version}" semver range. No need to upgrade`,
      );
      return null;
    }
    logger.error(
      `Dependency mismatch. Specified version "${newVersion}" is already installed in node_modules and it doesn't satisfy "${version}" semver range of your "react-native" dependency. Please re-install your dependencies`,
    );
    return null;
  }

  return newVersion;
};

const installDeps = async (newVersion, projectDir, patchSuccess) => {
  if (!patchSuccess) {
    logger.warn(
      'Continuing after failure. Most of the files are upgraded but you will need to deal with some conflicts manually',
    );
  }
  logger.info(
    `Installing react-native@${newVersion} and its peer dependencies...`,
  );
  const peerDeps = await getRNPeerDeps(newVersion);
  const pm = new PackageManager({projectDir});
  const deps = [
    `react-native@${newVersion}`,
    ...Object.keys(peerDeps).map(module => `${module}@${peerDeps[module]}`),
  ];
  await pm.install(deps, {silent: true});
  await execa('git', ['add', 'package.json']);
  try {
    await execa('git', ['add', 'yarn.lock']);
  } catch (error) {
    // ignore
  }
  try {
    await execa('git', ['add', 'package-lock.json']);
  } catch (error) {
    // ignore
  }
};

const applyPatch = async (
  currentVersion: string,
  newVersion: string,
  tmpPatchFile: string,
) => {
  let filesToExclude = ['package.json'];
  try {
    try {
      const excludes = filesToExclude.map(e => `--exclude=${e}`);
      await execa('git', [
        'apply',
        '--check',
        tmpPatchFile,
        ...excludes,
        '-p2',
        '--3way',
      ]);
      logger.info('Applying diff...');
    } catch (error) {
      filesToExclude = [
        ...filesToExclude,
        ...error.stderr
          .split('\n')
          .filter(x => x.includes('does not exist in index'))
          .map(x => x.replace(/^error: (.*): does not exist in index$/, '$1')),
      ].filter(Boolean);

      logger.info(`Applying diff (excluding: ${filesToExclude.join(', ')})...`);
    } finally {
      const excludes = filesToExclude.map(e => `--exclude=${e}`);
      await execa('git', ['apply', tmpPatchFile, ...excludes, '-p2', '--3way']);
    }
  } catch (error) {
    if (error.stderr) {
      logger.log(`${chalk.dim(error.stderr.trim())}`);
    }
    logger.error('Automatically applying diff failed');
    logger.info(
      `Here's the diff we tried to apply: ${rnDiffPurgeUrl}/compare/version/${currentVersion}...version/${newVersion}`,
    );
    logger.info(
      `You may find release notes helpful: https://github.com/facebook/react-native/releases/tag/v${newVersion}`,
    );
    return false;
  }
  return true;
};

/**
 * Upgrade application to a new version of React Native.
 */
async function upgrade(argv: Array<string>, ctx: ContextT, args: FlagsT) {
  if (args.legacy) {
    return legacyUpgrade.func(argv, ctx);
  }
  const rnDiffGitAddress =
    'https://github.com/react-native-community/rn-diff-purge.git';
  const tmpRemote = 'tmp-rn-diff-purge';
  const tmpPatchFile = 'tmp-upgrade-rn.patch';
  const projectDir = ctx.root;
  const {version: currentVersion} = require(path.join(
    projectDir,
    'node_modules/react-native/package.json',
  ));

  const newVersion = await getVersionToUpgradeTo(
    argv,
    currentVersion,
    projectDir,
  );

  if (!newVersion) {
    return;
  }

  const patch = await getPatch(currentVersion, newVersion, projectDir);

  if (patch === null) {
    return;
  }

  if (patch === '') {
    logger.info('Diff has no changes to apply, proceeding further');
    await installDeps(newVersion, projectDir);
    logger.success(
      `Upgraded React Native to v${newVersion} ðŸŽ‰. Now you can review and commit the changes`,
    );
    return;
  }
  let patchSuccess;

  try {
    fs.writeFileSync(tmpPatchFile, patch);
    await execa('git', ['remote', 'add', tmpRemote, rnDiffGitAddress]);
    await execa('git', ['fetch', '--no-tags', tmpRemote]);
    patchSuccess = await applyPatch(currentVersion, newVersion, tmpPatchFile);
    if (!patchSuccess) {
      return;
    }
  } catch (error) {
    throw new Error(error.stderr || error);
  } finally {
    try {
      fs.unlinkSync(tmpPatchFile);
    } catch (e) {
      // ignore
    }
    await installDeps(newVersion, projectDir, patchSuccess);
    logger.info('Running "git status" to check what changed...');
    await execa('git', ['status'], {stdio: 'inherit'});
    await execa('git', ['remote', 'remove', tmpRemote]);

    if (!patchSuccess) {
      logger.warn(
        'Please run "git diff" to review the conflicts and resolve them',
      );
      throw new Error(
        'Upgrade failed. Please see the messages above for details',
      );
    }
  }
  logger.success(
    `Upgraded React Native to v${newVersion} ðŸŽ‰. Now you can review and commit the changes`,
  );
}
const upgradeCommand = {
  name: 'upgrade [version]',
  description:
    "Upgrade your app's template files to the specified or latest npm version using `rn-diff-purge` project. Only valid semver versions are allowed.",
  func: upgrade,
  options: [
    {
      command: '--legacy',
      description:
        "Legacy implementation. Upgrade your app's template files to the latest version; run this after " +
        'updating the react-native version in your package.json and running npm install',
    },
  ],
};
export default upgradeCommand;
