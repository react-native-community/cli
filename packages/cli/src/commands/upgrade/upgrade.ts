import path from 'path';
import fs from 'fs';
import chalk from 'chalk';
import semver from 'semver';
import execa from 'execa';
import {Config} from '@react-native-community/cli-types';
import {logger, CLIError, fetch} from '@react-native-community/cli-tools';
import * as PackageManager from '../../tools/packageManager';
import installPods from '../../tools/installPods';

type UpgradeError = {message: string; stderr: string};

// https://react-native-community.github.io/upgrade-helper/?from=0.59.10&to=0.60.0-rc.3

type RepoNameType = 'react-native' | 'react-native-tvos';

const repos = {
  'react-native': {
    rawDiffUrl:
      'https://raw.githubusercontent.com/react-native-community/rn-diff-purge/diffs/diffs',
    webDiffUrl: 'https://react-native-community.github.io/upgrade-helper',
    dependencyName: 'react-native',
  },
  'react-native-tvos': {
    rawDiffUrl:
      'https://raw.githubusercontent.com/react-native-tvos/rn-diff-purge-tv/diffs/diffs',
    webDiffUrl: 'https://react-native-community.github.io/upgrade-helper',
    dependencyName: 'react-native@npm:react-native-tvos',
  },
};

const isConnected = (output: string): boolean => {
  // there is no reliable way of checking for internet connectivity, so we should just
  // read the output from npm (to check for connectivity errors) which is faster and relatively more reliable.
  return !output.includes('the host is inaccessible');
};

const checkForErrors = (output: string): void => {
  if (!output) {
    return;
  }
  if (!isConnected(output)) {
    throw new CLIError(
      'Upgrade failed. You do not seem to have an internet connection.',
    );
  }

  if (output.includes('npm ERR')) {
    throw new CLIError(`Upgrade failed with the following errors:\n${output}`);
  }

  if (output.includes('npm WARN')) {
    logger.warn(output);
  }
};

const getLatestRNVersion = async (repoName: RepoNameType): Promise<string> => {
  logger.info('No version passed. Fetching latest...');
  const {stdout, stderr} = await execa('npm', ['info', repoName, 'version']);
  checkForErrors(stderr);
  return stdout;
};

const getRNPeerDeps = async (
  version: string,
  repoName: RepoNameType,
): Promise<{[key: string]: string}> => {
  const {stdout, stderr} = await execa('npm', [
    'info',
    `${repoName}@${version}`,
    'peerDependencies',
    '--json',
  ]);
  checkForErrors(stderr);
  return JSON.parse(stdout);
};

const getPatch = async (
  currentVersion: string,
  newVersion: string,
  config: Config,
  repoName: RepoNameType,
) => {
  let patch;

  logger.info(`Fetching diff between v${currentVersion} and v${newVersion}...`);

  try {
    const {data} = await fetch(
      `${repos[repoName].rawDiffUrl}/${currentVersion}..${newVersion}.diff`,
    );

    patch = data;
  } catch (error) {
    logger.error((error as UpgradeError).message);
    logger.error(
      `Failed to fetch diff for react-native@${newVersion}. Maybe it's not released yet?`,
    );
    logger.info(
      `For available releases to diff see: ${chalk.underline.dim(
        'https://github.com/react-native-community/rn-diff-purge#diff-table-full-table-here',
      )}`,
    );
    return null;
  }

  let patchWithRenamedProjects = patch;

  Object.keys(config.project).forEach((platform) => {
    if (!config.project[platform]) {
      return;
    }
    if (platform === 'ios') {
      patchWithRenamedProjects = patchWithRenamedProjects.replace(
        new RegExp('RnDiffApp', 'g'),
        config.project[platform]!.projectName.replace('.xcodeproj', ''),
      );
    } else if (platform === 'android') {
      patchWithRenamedProjects = patchWithRenamedProjects
        .replace(
          new RegExp('com\\.rndiffapp', 'g'),
          config.project[platform]!.packageName,
        )
        .replace(
          new RegExp('com\\.rndiffapp'.split('.').join('/'), 'g'),
          config.project[platform]!.packageName.split('.').join('/'),
        );
    } else {
      logger.warn(
        `Unsupported platform: "${platform}". \`upgrade\` only supports iOS and Android.`,
      );
    }
  });

  return patchWithRenamedProjects;
};

const getVersionToUpgradeTo = async (
  argv: Array<string>,
  currentVersion: string,
  projectDir: string,
  repoName: RepoNameType,
) => {
  const argVersion = argv[0];
  const semverCoercedVersion = semver.coerce(argVersion);
  const newVersion = argVersion
    ? semver.valid(argVersion) ||
      (semverCoercedVersion ? semverCoercedVersion.version : null)
    : await getLatestRNVersion(repoName);

  if (!newVersion) {
    logger.error(
      `Provided version "${argv[0]}" is not allowed. Please pass a valid semver version`,
    );
    return null;
  }

  if (semver.gt(currentVersion, newVersion)) {
    logger.error(
      `Trying to upgrade from newer version "${currentVersion}" to older "${newVersion}"`,
    );
    return null;
  }
  if (semver.eq(currentVersion, newVersion)) {
    const {
      dependencies: {'react-native': version},
    } = require(path.join(projectDir, 'package.json'));

    const parsedVersion = version.split('@')[version.split('@').length - 1];

    if (semver.satisfies(newVersion, parsedVersion)) {
      logger.warn(
        `Specified version "${newVersion}" is already installed in node_modules and it satisfies "${parsedVersion}" semver range. No need to upgrade`,
      );
      return null;
    }
    logger.error(
      `Dependency mismatch. Specified version "${newVersion}" is already installed in node_modules and it doesn't satisfy "${parsedVersion}" semver range of your "react-native" dependency. Please re-install your dependencies`,
    );
    return null;
  }

  return newVersion;
};

const installDeps = async (
  root: string,
  newVersion: string,
  repoName: RepoNameType,
) => {
  logger.info(
    `Installing "react-native@${newVersion}" and its peer dependencies...`,
  );
  const peerDeps = await getRNPeerDeps(newVersion, repoName);
  const deps = [
    `${repos[repoName].dependencyName}@${newVersion}`,
    ...Object.keys(peerDeps).map((module) => `${module}@${peerDeps[module]}`),
  ];
  await PackageManager.install(deps, {
    silent: true,
    root,
  });
  await execa('git', ['add', 'package.json']);
  try {
    await execa('git', ['add', 'yarn.lock']);
  } catch (error) {
    // ignore
  }
  try {
    await execa('git', ['add', 'package-lock.json']);
  } catch (error) {
    // ignore
  }
};

const installCocoaPodsDeps = async (projectDir: string) => {
  if (process.platform === 'darwin') {
    try {
      logger.info(
        `Installing CocoaPods dependencies ${chalk.dim(
          '(this may take a few minutes)',
        )}`,
      );
      await installPods({
        directory: projectDir.split('/').pop() || '',
      });
    } catch (error) {
      if ((error as UpgradeError).stderr) {
        logger.debug(
          `"pod install" or "pod repo update" failed. Error output:\n${
            (error as UpgradeError).stderr
          }`,
        );
      }
      logger.error(
        'Installation of CocoaPods dependencies failed. Try to install them manually by running "pod install" in "ios" directory after finishing upgrade',
      );
    }
  }
};

const applyPatch = async (
  currentVersion: string,
  newVersion: string,
  tmpPatchFile: string,
  repoName: RepoNameType,
) => {
  const defaultExcludes = ['package.json'];
  let filesThatDontExist: Array<string> = [];
  let filesThatFailedToApply: Array<string> = [];

  const {stdout: relativePathFromRoot} = await execa('git', [
    'rev-parse',
    '--show-prefix',
  ]);
  try {
    try {
      const excludes = defaultExcludes.map(
        (e) => `--exclude=${path.join(relativePathFromRoot, e)}`,
      );
      await execa('git', [
        'apply',
        // According to git documentation, `--binary` flag is turned on by
        // default. However it's necessary when running `git apply --check` to
        // actually accept binary files, maybe a bug in git?
        '--binary',
        '--check',
        tmpPatchFile,
        ...excludes,
        '-p2',
        '--3way',
        `--directory=${relativePathFromRoot}`,
      ]);
      logger.info('Applying diff...');
    } catch (error) {
      const errorLines: Array<string> = (error as UpgradeError).stderr.split(
        '\n',
      );
      filesThatDontExist = [
        ...errorLines
          .filter((x) => x.includes('does not exist in index'))
          .map((x) =>
            x.replace(/^error: (.*): does not exist in index$/, '$1'),
          ),
      ].filter(Boolean);

      filesThatFailedToApply = errorLines
        .filter((x) => x.includes('patch does not apply'))
        .map((x) => x.replace(/^error: (.*): patch does not apply$/, '$1'))
        .filter(Boolean);

      logger.info('Applying diff...');
      logger.warn(
        `Excluding files that exist in the template, but not in your project:\n${filesThatDontExist
          .map((file) => `  - ${chalk.bold(file)}`)
          .join('\n')}`,
      );
      if (filesThatFailedToApply.length) {
        logger.error(
          `Excluding files that failed to apply the diff:\n${filesThatFailedToApply
            .map((file) => `  - ${chalk.bold(file)}`)
            .join(
              '\n',
            )}\nPlease make sure to check the actual changes after the upgrade command is finished.\nYou can find them in our Upgrade Helper web app: ${chalk.underline.dim(
            `${repos[repoName].webDiffUrl}/?from=${currentVersion}&to=${newVersion}`,
          )}`,
        );
      }
    } finally {
      const excludes = [
        ...defaultExcludes,
        ...filesThatDontExist,
        ...filesThatFailedToApply,
      ].map((e) => `--exclude=${path.join(relativePathFromRoot, e)}`);
      await execa('git', [
        'apply',
        tmpPatchFile,
        ...excludes,
        '-p2',
        '--3way',
        `--directory=${relativePathFromRoot}`,
      ]);
    }
  } catch (error) {
    if ((error as UpgradeError).stderr) {
      logger.debug(
        `"git apply" failed. Error output:\n${(error as UpgradeError).stderr}`,
      );
    }
    logger.error(
      'Automatically applying diff failed. We did our best to automatically upgrade as many files as possible',
    );
    return false;
  }
  return true;
};

/**
 * Upgrade application to a new version of React Native.
 */
async function upgrade(argv: Array<string>, ctx: Config) {
  const tmpPatchFile = 'tmp-upgrade-rn.patch';
  const projectDir = ctx.root;
  const {name: rnName, version: currentVersion} = require(path.join(
    projectDir,
    'node_modules/react-native/package.json',
  ));

  const repoName: RepoNameType =
    rnName === 'react-native-tvos' ? 'react-native-tvos' : 'react-native';

  const newVersion = await getVersionToUpgradeTo(
    argv,
    currentVersion,
    projectDir,
    repoName,
  );

  if (!newVersion) {
    return;
  }

  const patch = await getPatch(currentVersion, newVersion, ctx, repoName);

  if (patch === null) {
    return;
  }

  if (patch === '') {
    logger.info('Diff has no changes to apply, proceeding further');
    await installDeps(projectDir, newVersion, repoName);
    await installCocoaPodsDeps(projectDir);

    logger.success(
      `Upgraded React Native to v${newVersion} 🎉. Now you can review and commit the changes`,
    );
    return;
  }
  let patchSuccess;

  try {
    fs.writeFileSync(tmpPatchFile, patch);
    patchSuccess = await applyPatch(
      currentVersion,
      newVersion,
      tmpPatchFile,
      repoName,
    );
  } catch (error) {
    throw new Error((error as UpgradeError).stderr || (error as string));
  } finally {
    try {
      fs.unlinkSync(tmpPatchFile);
    } catch (e) {
      // ignore
    }
    const {stdout} = await execa('git', ['status', '-s']);
    if (!patchSuccess) {
      if (stdout) {
        logger.warn(
          'Continuing after failure. Some of the files are upgraded but you will need to deal with conflicts manually',
        );
        await installDeps(projectDir, newVersion, repoName);
        logger.info('Running "git status" to check what changed...');
        await execa('git', ['status'], {stdio: 'inherit'});
      } else {
        logger.error(
          'Patch failed to apply for unknown reason. Please fall back to manual way of upgrading',
        );
      }
    } else {
      await installDeps(projectDir, newVersion, repoName);
      await installCocoaPodsDeps(projectDir);
      logger.info('Running "git status" to check what changed...');
      await execa('git', ['status'], {stdio: 'inherit'});
    }
    if (!patchSuccess) {
      if (stdout) {
        logger.warn(
          'Please run "git diff" to review the conflicts and resolve them',
        );
      }
      if (process.platform === 'darwin') {
        logger.warn(
          'After resolving conflicts don\'t forget to run "pod install" inside "ios" directory',
        );
      }
      logger.info(`You may find these resources helpful:
• Release notes: ${chalk.underline.dim(
        `https://github.com/facebook/react-native/releases/tag/v${newVersion}`,
      )}
• Manual Upgrade Helper: ${chalk.underline.dim(
        `${repos[repoName].webDiffUrl}/?from=${currentVersion}&to=${newVersion}`,
      )}
• Git diff: ${chalk.underline.dim(
        `${repos[repoName].rawDiffUrl}/${currentVersion}..${newVersion}.diff`,
      )}`);

      throw new CLIError(
        'Upgrade failed. Please see the messages above for details',
      );
    }
  }
  logger.success(
    `Upgraded React Native to v${newVersion} 🎉. Now you can review and commit the changes`,
  );
}
const upgradeCommand = {
  name: 'upgrade [version]',
  description:
    "Upgrade your app's template files to the specified or latest npm version using `rn-diff-purge` project. Only valid semver versions are allowed.",
  func: upgrade,
};
export default upgradeCommand;
