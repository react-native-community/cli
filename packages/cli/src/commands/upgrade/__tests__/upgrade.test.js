// @flow
import execa from 'execa';
import path from 'path';
import fs from 'fs';
import snapshotDiff from 'snapshot-diff';
import upgrade from '../upgrade';
import {fetch} from '../helpers';
import logger from '../../../tools/logger';

jest.mock('https');
jest.mock('fs');
jest.mock('path');
jest.mock('execa', () => {
  const module = jest.fn((command, args) => {
    mockPushLog('$', 'execa', command, args);
    if (command === 'npm' && args[3] === '--json') {
      return Promise.resolve({
        stdout: '{"react": "16.6.3"}',
      });
    }
    return Promise.resolve({stdout: ''});
  });
  return module;
});
jest.mock(
  '/project/root/node_modules/react-native/package.json',
  () => ({name: 'react-native', version: '0.57.8'}),
  {virtual: true},
);
jest.mock(
  '/project/root/package.json',
  () => ({name: 'TestApp', dependencies: {'react-native': '^0.57.8'}}),
  {virtual: true},
);
jest.mock('../../../tools/PackageManager', () => ({
  install: args => {
    mockPushLog('$ yarn add', ...args);
  },
}));
jest.mock('../helpers', () => ({
  ...jest.requireActual('../helpers'),
  fetch: jest.fn(() => Promise.resolve('patch')),
}));
jest.mock('../../../tools/logger', () => ({
  info: jest.fn((...args) => mockPushLog('info', args)),
  error: jest.fn((...args) => mockPushLog('error', args)),
  warn: jest.fn((...args) => mockPushLog('warn', args)),
  success: jest.fn((...args) => mockPushLog('success', args)),
  log: jest.fn((...args) => mockPushLog(args)),
}));

const currentVersion = '0.57.8';
const newVersion = '0.58.4';
const olderVersion = '0.56.0';
const ctx = {
  root: '/project/root',
  reactNativePath: '',
};
const opts = {
  legacy: false,
};

const samplePatch = jest
  .requireActual('fs')
  .readFileSync(path.join(__dirname, './sample.patch'), 'utf8');

let logs = [];
const mockPushLog = (...args) =>
  logs.push(args.map(x => (Array.isArray(x) ? x.join(' ') : x)).join(' '));
const flushOutput = () => logs.join('\n');

beforeEach(() => {
  jest.clearAllMocks();
  // $FlowFixMe
  fs.writeFileSync = jest.fn(filename => mockPushLog('[fs] write', filename));
  // $FlowFixMe
  fs.unlinkSync = jest.fn((...args) => mockPushLog('[fs] unlink', args));
  logs = [];
});

test('uses latest version of react-native when none passed', async () => {
  await upgrade.func([], ctx, opts);
  expect(execa).toBeCalledWith('npm', ['info', 'react-native', 'version']);
});

test('errors when invalid version passed', async () => {
  await upgrade.func(['next'], ctx, opts);
  expect(logger.error).toBeCalledWith(
    'Provided version "next" is not allowed. Please pass a valid semver version',
  );
});

test('errors when older version passed', async () => {
  await upgrade.func([olderVersion], ctx, opts);
  expect(logger.error).toBeCalledWith(
    `Trying to upgrade from newer version "${currentVersion}" to older "${olderVersion}"`,
  );
});

test('warns when dependency upgrade version is in semver range', async () => {
  await upgrade.func([currentVersion], ctx, opts);
  expect(logger.warn).toBeCalledWith(
    `Specified version "${currentVersion}" is already installed in node_modules and it satisfies "^0.57.8" semver range. No need to upgrade`,
  );
});

test('fetches empty patch and installs deps', async () => {
  (fetch: any).mockImplementation(() => Promise.resolve(''));
  await upgrade.func([newVersion], ctx, opts);
  expect(flushOutput()).toMatchInlineSnapshot(`
"info Fetching diff between v0.57.8 and v0.58.4...
info Diff has no changes to apply, proceeding further
info Installing \\"react-native@0.58.4\\" and its peer dependencies...
$ execa npm info react-native@0.58.4 peerDependencies --json
$ yarn add react-native@0.58.4 react@16.6.3
$ execa git add package.json
$ execa git add yarn.lock
$ execa git add package-lock.json
success Upgraded React Native to v0.58.4 ðŸŽ‰. Now you can review and commit the changes"
`);
});

test('fetches regular patch, adds remote, applies patch, installs deps, removes remote,', async () => {
  (fetch: any).mockImplementation(() => Promise.resolve(samplePatch));
  await upgrade.func([newVersion], ctx, opts);
  expect(flushOutput()).toMatchInlineSnapshot(`
"info Fetching diff between v0.57.8 and v0.58.4...
[fs] write tmp-upgrade-rn.patch
$ execa git remote add tmp-rn-diff-purge https://github.com/react-native-community/rn-diff-purge.git
$ execa git fetch --no-tags tmp-rn-diff-purge
$ execa git apply --check tmp-upgrade-rn.patch --exclude=package.json -p2 --3way
info Applying diff...
$ execa git apply tmp-upgrade-rn.patch --exclude=package.json -p2 --3way
[fs] unlink tmp-upgrade-rn.patch
$ execa git status -s
info Installing \\"react-native@0.58.4\\" and its peer dependencies...
$ execa npm info react-native@0.58.4 peerDependencies --json
$ yarn add react-native@0.58.4 react@16.6.3
$ execa git add package.json
$ execa git add yarn.lock
$ execa git add package-lock.json
info Running \\"git status\\" to check what changed...
$ execa git status
$ execa git remote remove tmp-rn-diff-purge
success Upgraded React Native to v0.58.4 ðŸŽ‰. Now you can review and commit the changes"
`);
  expect(
    snapshotDiff(samplePatch, fs.writeFileSync.mock.calls[0][1], {
      contextLines: 1,
    }),
  ).toMatchSnapshot('RnDiffApp is replaced with app name (TestApp)');
});
test('cleans up if patching fails,', async () => {
  (fetch: any).mockImplementation(() => Promise.resolve(samplePatch));
  (execa: any).mockImplementation((command, args) => {
    mockPushLog('$', 'execa', command, args);
    if (command === 'npm' && args[3] === '--json') {
      return Promise.resolve({
        stdout: '{"react": "16.6.3"}',
      });
    }
    if (command === 'git' && args[0] === 'apply') {
      return Promise.reject({
        code: 1,
        stderr: 'error: .flowconfig: does not exist in index\n',
      });
    }
    return Promise.resolve({stdout: ''});
  });
  try {
    await upgrade.func([newVersion], ctx, opts);
  } catch (error) {
    expect(error.message).toBe(
      'Upgrade failed. Please see the messages above for details',
    );
  }
  expect(flushOutput()).toMatchInlineSnapshot(`
"info Fetching diff between v0.57.8 and v0.58.4...
[fs] write tmp-upgrade-rn.patch
$ execa git remote add tmp-rn-diff-purge https://github.com/react-native-community/rn-diff-purge.git
$ execa git fetch --no-tags tmp-rn-diff-purge
$ execa git apply --check tmp-upgrade-rn.patch --exclude=package.json -p2 --3way
info Applying diff (excluding: package.json, .flowconfig)...
$ execa git apply tmp-upgrade-rn.patch --exclude=package.json --exclude=.flowconfig -p2 --3way
[2merror: .flowconfig: does not exist in index[22m
error Automatically applying diff failed
[fs] unlink tmp-upgrade-rn.patch
$ execa git status -s
error Patch failed to apply for unknown reason. Please fall back to manual way of upgrading
$ execa git remote remove tmp-rn-diff-purge
info You may find these resources helpful:
â€¢ Release notes: [4m[2mhttps://github.com/facebook/react-native/releases/tag/v0.58.4[22m[24m
â€¢ Comparison between versions: [4m[2mhttps://github.com/react-native-community/rn-diff-purge/compare/version/0.57.8..version/0.58.4[22m[24m
â€¢ Git diff: [4m[2mhttps://github.com/react-native-community/rn-diff-purge/compare/version/0.57.8..version/0.58.4.diff[22m[24m"
`);
});
