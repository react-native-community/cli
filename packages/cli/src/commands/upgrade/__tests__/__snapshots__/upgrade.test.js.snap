// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fetches regular patch, adds remote, applies patch, installs deps, removes remote,: RnDiffApp is replaced with app name (TestApp) 1`] = `
"Snapshot Diff:
- First value
+ Second value

@@ -1,5 +1,5 @@
- diff --git a/RnDiffApp/android/build.gradle b/RnDiffApp/android/build.gradle
+ diff --git a/TestApp/android/build.gradle b/TestApp/android/build.gradle
  index 85d8f2f8..a1e80854 100644
- --- a/RnDiffApp/android/build.gradle
- +++ b/RnDiffApp/android/build.gradle
+ --- a/TestApp/android/build.gradle
+ +++ b/TestApp/android/build.gradle
  @@ -9,8 +9,8 @@ buildscript {
@@ -28,6 +28,6 @@
  
- diff --git a/RnDiffApp/package.json b/RnDiffApp/package.json
+ diff --git a/TestApp/package.json b/TestApp/package.json
  index 4e617645..c82829bd 100644
- --- a/RnDiffApp/package.json
- +++ b/RnDiffApp/package.json
+ --- a/TestApp/package.json
+ +++ b/TestApp/package.json
  @@ -7,14 +7,14 @@"
`;
